### ISSUE TITLE
Latency reporting

### ISSUE THREAD

**Issue by ioctlLR** [naudio/NAudio#601]
When I originally suggested the IWavePosition interface, it was so I could sync a visualization with audio being played.  I realize that was not the correct approach, so I'd like to suggest a new one:
```
public interface IReportLatency
{
    /// <summary>
    /// Gets the average amount of time this instance takes from receiving a sample to the sample being output by the hardware.  Assumes uninterrupted playback.
    /// </summary>
    TimeSpan AverageLatency { get; }

    /// <summary>
    /// Gets the amount of time that has passed since the sample currently being output by the hardware was buffered.  Assumes uninterrupted playback.
    /// </summary>
    TimeSpan CurrentLatency { get; }
}
```
I'm not stuck on the verbiage of the summaries or the naming of the interface, but there it is.

Implementation-wise, this should be pretty straightforward on most outputs: AverageLatency would be the duration of all but half the last buffer, and CurrentLatency just requires tracking when each buffer was last filled.  For certain other outputs (e.g., a hypothetical PulseAudioOut) we'd also have to ask the driver for further latency info, but even that shouldn't be too bad.

Thoughts?

**Comment by markheath** [comment]
Yes, this would be a useful capability, especially if we can return meaningful values for most supported output devices. Would be happy to accept a PR for this

**Comment by ioctlLR** [comment]
FYI, I started on this but haven't been able to spend much time on it...  I'll get there, eventually. :)

**Comment by markheath** [comment]
Been a long time but have had a quick look into whether this is possible with a slightly renamed interface - `IWaveLatency`

If you have time, check out the changes here https://github.com/naudio/NAudio/compare/feature/wave-latency-interface?expand=1 and let me know what you think.

**Comment by markheath** [comment]
Also decided to see if the same interface could be used against recording - seems like its possible and have a PR for easier reviewing. Not 100% decided on this yet, so will leave it a short while before completing to give people the chance to give feedback

**Comment by ioctlLR** [comment]
So I _had_ an implementation of this... But it's on a drive I currently can't access so I'll have to wait for the part to arrive (Monday?) so I can see what I was intending back then.
As I recall the purpose was to sync a Winamp-style spectrum visual. I never got happy with the level of complexity in my implementation though, so it never got uploaded anywhere. Yours looks a lot simpler than what I recall of mine.

**Comment by ioctlLR** [comment]
OK, so after extensive searching... I still can't find it. It was about four or five laptop rebuilds ago, so I'm not shocked.

The linked PR does pretty much what I remember looking for, so I vote to merge and close this issue whenever you are happy with it.

**Comment by markheath** [comment]
Thanks, no worries. Will look to get this merged for inclusion in NAudio 3
